apps/greencheck/microsoft_bucket_updater.py

```diff
@@ -0,0 +1,116 @@
+import requests
+
+from datetime import datetime, timedelta
+from . import object_storage
+
+
+class MicrosoftBucketUpdater:
+    def __init__(self):
+        self.ms_endpoint_url_prefix = """https://download.microsoft.com/download/7/1/D/
+        71D86715-5596-4529-9B13-DA13A5DE5B63/ServiceTags_Public_"""
+        self.bucket_url_prefix = "data-imports/microsoft-network-ranges-"
+        self.file_extension = ".json"
+
+    def search_for_dataset(self) -> dict:
+        """
+        Searching for an endpoint from Microsoft that returns a dataset containing
+        active network (IP/ASN) ranges.
+
+        Searching is being done by doing a HTTP request to a specific address and
+        trying out different dates at the ending of this address.
+
+        The reason for searching backwards in time is to ensure to always get the
+        latest version of the dataset first, in the case of having multiple endpoints
+        active at once.
+
+        Return: dict (representing a dataset from Microsoft containing active IP ranges)
+        """
+        # Set a range to start searching from:
+        # - pivot : iterable date, starting with today
+        # - until : date to search up to (difference between today and last updated)
+        date_pivot = datetime.now()
+        duration_between_last_updated = self.retrieve_delta_last_updated()
```

Comment:
@mrchrisadams we've talked about this on our call today.
We talked about searching for a endpoint that returns a dataset from today until a certain point.

Firstly we discussed that we can search all the way back to the first ever updated. Secondly we discussed that it was simply one week. But it seems to be easier than that. As you can see, the function _retrieve_delta_last_update_ simply retrieves the time between now and the date of the last updated file in our object storage. This makes it much more reliable and can therefore dynamically adjust in relation to Microsoft's updating schedule.

For example, if the last updated was a week ago, then it will search for 7 days. Simply because there are 7 days between today and the one that was last updated. If this was one month ago, it will search the whole month until it finds a working endpoint. If it finds something, then it will add it to the object storage and that will be our "last updated" file.